pki/blacklist/blacklist.go

```diff
@@ -0,0 +1,244 @@
+	}
+
+	// Return the new blacklist and a nil error
+	return &blacklistImpl{
+		trustedKey: key,
+		url:        config.URL,
+	}, nil
+}
+
+// ValidateCert checks for the issuer and serialNumber in the blacklist, returning nil when the cert is permitted
+// or an error when the cert is blacklisted or the blacklist cannot be retrieved
+func (b *blacklistImpl) ValidateCert(cert *x509.Certificate) error {
+	// Blacklists with an empty URL are a NOP
+	if b.URL() == "" {
+		return nil
+	}
+
+	// Extract the necessary information from the certificate in order to process the blacklist entry
+	issuer := cert.Issuer.String()
+	serialNumber := cert.SerialNumber.String()
+	thumbprint := certKeyJWKThumbprint(cert)	
+
+	// If the blacklist has not yet been downloaded, do so now
+	if b.lastUpdated.IsZero() {
+		// Trigger an update of the blacklist
+		if err := b.Update(); err != nil {
+			// If the blacklist download failed then log a message about it
+			logger().WithError(err).
+				WithField("Issuer", issuer).
+				WithField("S/N", serialNumber).
+				Error("cert validation failed because the blacklist cannot be downloaded")
+
+			// Return an error indicating the blacklist cannot be retrieved
+			return ErrBlacklistMissing
+		}
+	}
+
+	// Fetch the entries
+	entriesPtr := b.entries.Load()
+
+	// The entries pointer should not be empty because of the lastUpdated check above
+	if entriesPtr == nil {
+		// If the entries still cannot be fetched then something is not right so return an error
+		return ErrBlacklistMissing
+	}
+
+	// Check each entry in the blacklist for matches
+	for _, entry := range *entriesPtr {
+		// Check for this issuer and serial number combination
+		if entry.Issuer == issuer && entry.SerialNumber == serialNumber && entry.JWKThumbprint == thumbprint {
+			// Return an error indicating the certificate has been blacklisted
+			return ErrCertBlacklisted
+		}
+	}
+
+	// Return a nil error as the certificate hasn't been blacklisted
+	return nil
+}
+
+func (b *blacklistImpl) LastUpdated() time.Time {
+	return b.lastUpdated
+}
+
+func (b *blacklistImpl) URL() string {
+	return b.url
+}
+
+// update downloads the blacklist, and updates the in-memory representation
+func (b *blacklistImpl) Update() error {
+	// Updating a blacklist with a URL is a NOP
+	if b.URL() == "" {
+		return nil
+	}
+
+	// Download the blacklist
+	bytes, err := b.download()
+	if err != nil {
+		logger().WithError(err).
+			WithField("URl", b.url).
+			Warn("certiciate blacklist cannot be downloaded")
+		return err
+	}
+
+	// Check the signature of the blacklist
+	payload, err := jws.Verify(bytes, jwa.EdDSA, b.trustedKey)
+	if err != nil {
+		return fmt.Errorf("failed to verify blacklist signature: %w", err)
+	}
+
+	// Parse the JSON of the payload
+	var entries []blacklistEntry
+	if err = json.Unmarshal(payload, &entries); err != nil {
+		return fmt.Errorf("failed to parse blacklist JSON: %w", err)
+	}
+
+	// Store the new blacklist entries
+	b.entries.Store(&entries)
```

Comment:
I'd like to have a test validating this for when the blocklist is empty. (unless I missed it).